Catching Unknown status of node when cluster becomes disconnected

paigerube14 opened the pull request:
### Description
This fix adds in exception catches when performing node reboot and the cluster gets disconnected. Also added in a timeout to the get_node_status function to help with this

### Fixes
https://github.com/cloud-bulldozer/kraken/issues/216

paigerube14 committed:
Updating unknown status for when cluster becomes disconnected

chaitanyaenr commented on kraken/kubernetes/client.py:
```diff
@@ -274,15 +274,16 @@ def get_job_status(name, namespace="default"):
+def get_node_status(node, timeout=60):
     try:
-        node_info = cli.read_node_status(node, pretty=True)
+        node_info = cli.read_node_status(node, pretty=True, _request_timeout=timeout)
```
The parameter must be named as _request_timeout in Kubernetes python client. Just want to make sure that's the case. Thoughts?

paigerube14 replied:
Yes I think it is properly named 


```
 def read_node_status(self, name, **kwargs):  # noqa: E501
        """read_node_status  # noqa: E501

        read status of the specified Node  # noqa: E501
        This method makes a synchronous HTTP request by default. To make an
        asynchronous HTTP request, please pass async_req=True
        >>> thread = api.read_node_status(name, async_req=True)
        >>> result = thread.get()

        :param async_req bool: execute request asynchronously
        :param str name: name of the Node (required)
        :param str pretty: If 'true', then the output is pretty printed.
        :param _preload_content: if False, the urllib3.HTTPResponse object will
                                 be returned without reading/decoding response
                                 data. Default is True.
        :param _request_timeout: timeout setting for this request. If one
                                 number provided, it will be total request
                                 timeout. It can also be a pair (tuple) of
                                 (connection, read) timeouts.
        :return: V1Node
                 If the method is called asynchronously,
                 returns the request thread.
        """
```

chaitanyaenr reviewed (approved):
LGTM

sandrobonazzola commented:
"This fix adds in exception catches when performing node reboot and the cluster gets disconnected. Also added in a timeout to the get_node_status function to help with this"

The "Also added..." part suggests this could have been split in two separate PRs, making review easier.

sandrobonazzola reviewed (changes requested):
Comments inline

paigerube14 commented:
> 

The also part I just had to add in to be able to properly get a status back if the cluster becomes disconnected. It was part of the update and I think should still be in this PR

paigerube14 commented on kraken/node_actions/common_node_functions.py:
```diff
@@ -32,17 +32,26 @@ def get_node(node_name, label_selector, instance_kill_count):
 # Wait till node status becomes Ready
 def wait_for_ready_status(node, timeout):
-    runcommand.invoke("kubectl wait --for=condition=Ready " "node/" + node + " --timeout=" + str(timeout) + "s")
+    runcommand.invoke("kubectl wait --for=condition=Ready node/" + node + " --timeout=" + str(timeout) + "s")
```
Had already created an issue to update this in the near future: https://github.com/cloud-bulldozer/kraken/issues/218

paigerube14 commented:
> "This fix adds in exception catches when performing node reboot and the cluster gets disconnected. Also added in a timeout to the get_node_status function to help with this"
> 
> The "Also added..." part suggests this could have been split in two separate PRs, making review easier.

This is also part of the fix I implemented to be able to properly get a status back if the cluster becomes disconnected. It was part of the update and I think should still be in this PR

yogananth-subramanian commented on kraken/node_actions/common_node_functions.py:
```diff
@@ -32,17 +32,26 @@ def get_node(node_name, label_selector, instance_kill_count):
+        node_status = kubecli.get_node_status(node, timeout)
+        logging.info("node status " + str(node_status))
+        if node_status is not None and node_status != "Unknown":
+            logging.error("Node condition status isn't Unknown after %s seconds" % str(timeout))
```
Should it be an Exception instead of just logging it as "isn't unknown" like before.
`raise Exception("Node condition status isn't Unknown")`

yogananth-subramanian commented on kraken/node_actions/common_node_functions.py:
```diff
@@ -32,17 +32,26 @@ def get_node(node_name, label_selector, instance_kill_count):
+        if node_status is not None and node_status != "Unknown":
+            logging.error("Node condition status isn't Unknown after %s seconds" % str(timeout))
+    except Exception:
+        logging.error("Encountered error while getting node status, for more than %s seconds" % str(timeout))
```
`ApiException` is already handled in get_node_status and any other exception that was raised by `wait_for_unknown_status` in the old was handled by the calling function, so is there a need to handle the exception here?

yogananth-subramanian commented on kraken/node_actions/common_node_functions.py:
```diff
@@ -32,17 +32,23 @@ def get_node(node_name, label_selector, instance_kill_count):
+    time.sleep(3)
+    node_status = kubecli.get_node_status(node, timeout)
+    logging.info("node status " + str(node_status))
+    if node_status is not None and node_status != "Unknown":
```
/lgtm
@paigerube14  If you could just rebase the patch, it would be ready to be merged.

paigerube14 replied:
@yogananth-subramanian done!
